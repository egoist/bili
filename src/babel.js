const env = process.env.BABEL_ENV || process.env.NODE_ENV

export default (
  ctx,
  { jsx, buble, target, objectAssign, env: envOption } = {}
) => {
  jsx = jsx || 'react'

  let presets = []
  let plugins = []

  if (jsx === 'vue') {
    plugins.push(require.resolve('babel-plugin-transform-vue-jsx'))
  } else if (jsx === 'react') {
    plugins.push(require.resolve('@babel/plugin-transform-react-jsx'))
  } else if (typeof jsx === 'string') {
    plugins.push([
      require.resolve('@babel/plugin-transform-react-jsx'),
      { pragma: jsx }
    ])
  }

  plugins.push(
    [
      require.resolve('fast-async'),
      {
        spec: true
      }
    ],
    require.resolve('@babel/plugin-transform-flow-strip-types'),
    [
      require.resolve('@babel/plugin-proposal-object-rest-spread'),
      {
        useBuiltIns: true,
        loose: true
      }
    ],
    [
      require.resolve('babel-plugin-alter-object-assign'),
      {
        objectAssign
      }
    ]
  )

  if (buble) {
    return {
      presets,
      plugins
    }
  }

  presets = [
    ...presets,
    env === 'test' ?
      [
        require('@babel/preset-env').default,
        {
          modules: false,
          targets: {
            node: 'current'
          },
          ...envOption
        }
      ] :
      [
        require('@babel/preset-env').default,
        {
          // Never polyfill something like `Promise` `Proxy`
          // Since we're building a library instead of an app
          // You should not include polyfill in your lib anyways
          useBuiltIns: false,
          modules: false,
          targets:
              target && target.startsWith('node') ?
                {
                  node: target.split(':')[1] || 6
                } :
                {
                  ie: 9
                },
          exclude: ['transform-regenerator', 'transform-async-to-generator'],
          ...envOption
        }
      ]
  ]

  plugins = [
    ...plugins,
    require.resolve('@babel/plugin-proposal-class-properties')
  ]

  return {
    presets,
    plugins
  }
}
