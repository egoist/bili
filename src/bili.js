import * as rollup from 'rollup'
import watch from 'rollup-watch'
import switchy from 'switchy'
import chalk from 'chalk'
import fancyLog from 'fancy-log'
import merge from 'lodash.merge'
import getRollupOptions from './get-rollup-options'
import getConfig from './get-config'

function log(type, msg, color) {
  if (!color) {
    fancyLog(`${type} ${msg}`)
    return
  }
  fancyLog(`${color(type)} ${msg}`)
}

export default function(options = {}) {
  const userConfig = getConfig(options.config)

  options = merge(
    {
      entry: './src/index.js',
      format: ['cjs'],
      outDir: './dist',
      name: 'index'
    },
    userConfig,
    options
  )

  let formats = options.format

  if (typeof formats === 'string') {
    if (formats === 'all') {
      formats = ['cjs', 'umd', 'es']
    } else if (formats.indexOf(',') > -1) {
      formats = formats.split(',').map(v => v.trim())
    } else {
      formats = [formats]
    }
  } else if (!Array.isArray(formats)) {
    throw new TypeError('Expect "format" to be a string or Array')
  }

  if (options.compress) {
    formats.push('umd-compress')
  }

  return Promise.all(
    formats.map(format => {
      const rollupOptions = getRollupOptions(options, format)
      if (options.watch) {
        let init
        return new Promise(resolve => {
          const watcher = watch(rollup, rollupOptions)
          watcher.on('event', event => {
            switchy({
              STARTING() {
                log(format, 'starting', chalk.white.bgBlue)
                if (!init) {
                  init = true
                  return resolve()
                }
              },
              BUILD_START() {},
              BUILD_END() {
                log(format, 'bundled successfully', chalk.black.bgGreen)
              },
              ERROR() {
                const error = event.error
                log(format, '', chalk.white.bgRed)
                if (error.snippet) {
                  console.error(chalk.red(`---\n${error.snippet}\n---`))
                }
                console.error(error.stack)
              },
              default() {
                console.error('unknown event', event)
              }
            })(event.code)
          })
        })
      }
      return rollup.rollup(rollupOptions).then(bundle => {
        return bundle.write(rollupOptions)
      })
    })
  )
}
