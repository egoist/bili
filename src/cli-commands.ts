import path from 'path'
import fs from 'fs'
import JoyCon from 'joycon'
import logger from './logger'

type fields = 'main' | 'module' | 'cdn' | 'unpkg' | 'jsdeliver'

function addProp(obj: any, prop: fields, val: string, fields: fields[]) {
  return Object.keys(obj).reduce((acc: any, item: string) => {
    if (item === 'main') {
      prop !== 'main' && (acc[item] = obj[item])
      acc[prop] = val
      fields.push(prop)
    } else {
      acc[item] = obj[item]
    }
    return acc
  }, {})
}

export const updatePackage = (
  outputConfig: any,
  rollupOutput: any,
  rootDir: string
): void => {
  const updatedFields: fields[] = []
  const pkgLoader = new JoyCon({
    stopDir: path.dirname(process.cwd())
  })

  const files = rollupOutput.output.map((output: any) => {
    return {
      name: output.name,
      fileName: output.fileName,
      dir: path.relative(rootDir, outputConfig.dir),
      format: outputConfig.format,
      outputDir:
        path.relative(rootDir, outputConfig.dir) + '/' + output.fileName,
      isChunk: !output.isEntry
    }
  })

  const pkg = pkgLoader.loadSync({
    files: ['package.json'],
    cwd: rootDir
  })

  const check = (file: any, format: string) => {
    return (
      file.format === format &&
      file.isChunk === false &&
      ['main', 'index'].includes(file.name)
    )
  }

  for (const file of files) {
    if (check(file, 'cjs'))
      pkg.data = addProp(pkg.data, 'main', file.outputDir, updatedFields)
    if (check(file, 'esm'))
      pkg.data = addProp(pkg.data, 'module', file.outputDir, updatedFields)
    if (check(file, 'umd')) {
      pkg.data = addProp(pkg.data, 'cdn', file.outputDir, updatedFields)
      pkg.data = addProp(pkg.data, 'unpkg', file.outputDir, updatedFields)
      pkg.data = addProp(pkg.data, 'jsdeliver', file.outputDir, updatedFields)
    }
  }

  fs.writeFileSync(
    `${rootDir}/package.json`,
    JSON.stringify(pkg.data, null, 2),
    'utf8'
  )
  logger.success(
    `Updated package.json ${updatedFields.join(',')} ${
      updatedFields.length === 1 ? 'field' : 'fields'
    } \n`
  )
}
