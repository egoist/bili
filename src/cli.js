#!/usr/bin/env node
import cac from 'cac'
import Bili from '.'

const cli = cac()

cli
  .command('*', 'Bundle library', async (input, flags) => {
    const options = {
      input,
      ...flags
    }
    if (options.debug) {
      options.logLevel = 4
    } else if (options.quiet) {
      options.logLevel = 1
    }
    return Bili.write(options)
  })
  .option('watch', {
    desc: 'Build and watch files',
    alias: 'w'
  })
  .option('outDir', {
    desc: 'Output directory',
    alias: ['o', 'd']
  })
  .option('moduleName', {
    desc: 'Module name for UMD/IIFE bundle'
  })
  .option('format', {
    desc: 'Output format'
  })
  .option('filename', {
    desc: 'Output filename, default: "[name][suffix].js"'
  })
  .option('name', {
    desc: 'Set the [name] part of "filename"'
  })
  .option('plugin', {
    desc: 'Add custom Rollup plugins'
  })
  .option('jsx', {
    desc: 'Switch JSX syntax, eg: "vue" "h"'
  })
  .option('global', {
    desc:
      'Mark module as global variable. Any module IDs defined here are added to external',
    alias: 'g'
  })
  .option('alias', {
    desc: 'Set option for rollup-plugin-alias'
  })
  .option('replace', {
    desc: 'Set option for rollup-plugin-replace'
  })
  .option('env', {
    desc: 'Replace env variables in your bundle',
    alias: 'e'
  })
  .option('inline', {
    desc: 'Inline external node modules'
  })
  .option('banner', {
    desc:
      'Content to insert to the top of bundle file, boolean or string or object'
  })
  .option('debug', {
    desc: 'Show debug logs'
  })
  .option('quiet', {
    desc: 'Show error logs only'
  })
  .option('no-logUpdate', {
    desc: 'Disable logUpdate'
  })

cli.parse()
