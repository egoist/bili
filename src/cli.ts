#!/usr/bin/env node
import 'v8-compile-cache'
import { cac } from 'cac'
import { version } from '../package.json'

if (process.env.BILI_LOCAL_PROFILE) {
  const requireSoSlow = require('require-so-slow')
  process.on('exit', () => {
    requireSoSlow.write('require-trace.trace')
  })
}

const cli = cac('bili')

cli
  .command('[...input]', 'Bundle input files', {
    ignoreOptionDefaultValue: true,
  })
  .option('-w, --watch', 'Watch files')
  .option(
    '--format <format>',
    'Output format (cjs | umd | es  | iife), can be used multiple times'
  )
  .option('--input.* [file]', 'An object mapping names to entry points')
  .option('-d, --out-dir <outDir>', 'Output directory', { default: 'dist' })
  .option('--root-dir <rootDir>', 'The root directory to resolve files from')
  .option('--file-name <name>', 'Set the file name for output files')
  .option('--module-name <name>', 'Set the module name for umd bundle')
  .option('--update-pkg', 'Updates your package.json entry fields')
  .option('--env.* [value]', 'Replace env variables')
  .option('--plugin, --plugins.* [options]', 'Use a plugin')
  .option(
    '--global.* [path]',
    'id:moduleName pair for external imports in umd/iife bundles'
  )
  .option('--no-extract-css', 'Do not extract CSS files')
  .option('--bundle-node-modules', 'Include node modules in your bundle')
  .option('--minify', 'Minify output files')
  .option('--external <id>', 'Mark a module id as external', {
    type: [],
  })
  .option('-t, --target <target>', 'Output target', { default: 'node' })
  .option('-c, --config <file>', 'Use a custom config file')
  .option('--minimal', 'Generate minimal output whenever possible')
  .option('--no-babelrc', 'Disable .babelrc file')
  .option('--banner', 'Add banner with pkg info to the bundle')
  .option(
    '--no-map',
    'Disable source maps, enabled by default for minified bundles'
  )
  .option('--map-exclude-sources', 'Exclude source code in source maps')
  .option('--no-async-pro, --no-async-to-promises', 'Leave async/await as is')
  .option('--concurrent', 'Build concurrently')
  .option('--verbose', 'Show verbose logs')
  .option('--quiet', 'Show minimal logs')
  .option('--stack-trace', 'Show stack trace for bundle errors')
  .example((bin) => `  ${bin} --format cjs --format esm`)
  .example((bin) => `  ${bin} src/index.js,src/cli.ts`)
  .example((bin) => `  ${bin} --input.index src/foo.ts`)
  .action(async (input, options) => {
    const { Bundler } = await import('./')
    const rootDir = options.rootDir || '.'
    const bundler = new Bundler(
      {
        input: options.input || (input.length === 0 ? undefined : input),
        output: {
          format: options.format,
          dir: options.outDir,
          moduleName: options.moduleName,
          fileName: options.fileName,
          minify: options.minify,
          extractCSS: options.extractCss,
          sourceMap: options.map,
          sourceMapExcludeSources: options.mapExcludeSources,
          target: options.target,
        },
        bundleNodeModules: options.bundleNodeModules,
        env: options.env,
        plugins: options.plugins,
        externals: options.external,
        globals: options.global,
        banner: options.banner,
        updatePkg: options.updatePkg,
        babel: {
          asyncToPromises: options.asyncToPromises,
          minimal: options.minimal,
          babelrc: options.babelrc,
        },
      },
      {
        logLevel: options.verbose
          ? 'verbose'
          : options.quiet
          ? 'quiet'
          : undefined,
        stackTrace: options.stackTrace,
        configFile: options.config,
        rootDir,
      }
    )
    await bundler
      .run({
        write: true,
        watch: options.watch,
        concurrent: options.concurrent,
      })
      .catch((err: any) => {
        bundler.handleError(err)
        process.exit(1)
      })
  })

cli.version(version)
cli.help()

cli.parse()

process.on('unhandledRejection', (err) => {
  console.error(err)
  process.exit(1)
})
