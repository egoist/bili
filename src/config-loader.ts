import fs from 'fs'
import JoyCon from 'joycon'
import path from 'path'
import requireFromString from 'require-from-string'

import logger from './logger'
import { Config } from './types'

const configLoader = new JoyCon({
  stopDir: path.dirname(process.cwd()),
})

configLoader.addLoader({
  test: /\.[jt]s$/,
  loadSync(id) {
    const content = require('@babel/core').transform(
      fs.readFileSync(id, 'utf8'),
      {
        babelrc: false,
        configFile: false,
        filename: id,
        presets: [
          [
            require('@babel/preset-env'),
            {
              targets: {
                node: 'current',
              },
            },
          ],
          id.endsWith('.ts') && require('@babel/preset-typescript'),
        ].filter(Boolean),
      }
    )
    const m = requireFromString(content && content.code ? content.code : '', id)
    return m.default || m
  },
})

function loadConfig(
  rootDir: string,
  configFile?: string | boolean
): {
  fileConfig: Config
  configPath?: string
} {
  let configPath: string | undefined

  const fileConfig =
    configFile === false
      ? {}
      : configLoader.loadSync({
          files:
            typeof configFile === 'string'
              ? [configFile]
              : [
                  'bili.config.js',
                  'bili.config.ts',
                  '.bilirc.js',
                  '.bilirc.ts',
                  'package.json',
                ],
          cwd: rootDir,
          packageKey: 'bili',
        })

  if (fileConfig.path) {
    logger.debug(`Using config file:`, fileConfig.path)
    configPath = fileConfig.path
  }

  return {
    fileConfig: fileConfig.data || {},
    configPath,
  }
}

export { loadConfig }
export default configLoader
