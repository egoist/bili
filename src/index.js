/* eslint-disable unicorn/no-process-exit */
import cac from 'cac'
import chalk from 'chalk'
import update from 'update-notifier'
import bili from './bili'

const cli = cac()

update({ pkg: cli.pkg }).notify()

cli
  .option('config, c', 'Path to config file')
  .option('watch, w', 'Run in watch mode')
  .option('name, n', 'The filename of output file, no extension')
  .option('out-dir, d', 'The output directory')
  .option('format', 'Bundle format, array of string')
  .option('module-name', 'The module name for UMD builds')
  .option('map', 'Generate sourcemap, boolean or `inline`')
  .option('compress', 'Generate a UMD bundle and compress it with sourcemaps')
  .option('skip', 'Exclude specfic modules in node_modules dir from bundled file')
  .option('es-module', 'Respect `jsnext:main` and `module` field in package.json')
  .option('browser', 'Respect `browser` field in package.json')
  .option('alias', 'Set option for rollup-plugin-alias')
  .option('replace', 'Set option for rollup-plugin-replace')
  .option('flow', 'Remove flow type annotations')
  .option('exports', 'Specific what export mode to use, `default` or `named`')
  .option('resolve', 'Resolve external dependencies')
  .option('banner', 'Content to insert at top of bundle, boolean or string or object')

cli.command('*', 'Bundle with bili', (input, flags) => {
  const options = Object.assign({
    entry: input[0]
  }, flags)

  return bili(options).catch(err => {
    console.log(err)
    if (err.snippet) {
      console.error(chalk.red(`---\n${err.snippet}\n---`))
    }
    console.error(err.message)
    process.exit(1) // eslint-disable-line xo/no-process-exit
  })
})

cli.parse()
