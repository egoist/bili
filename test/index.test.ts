import path from 'path'
import { Bundler, Config, BundleConfig, runBundler } from '../src'

process.env.BABEL_ENV = 'anything-not-test'

function fixture(...args: string[]) {
  return path.join(__dirname, 'fixtures', ...args)
}

function snapshot(
  {
    title,
    input,
    cwd,
  }: { title: string; input: string | string[]; cwd?: string },
  cliConfig?: BundleConfig,
  fileConfig: Config = {}
) {
  test(title, async () => {
    const bundler = await runBundler(
      {
        input,
        ...cliConfig,
      },
      fileConfig,
      {
        logLevel: 'quiet',
        rootDir: cwd,
      },
      {}
    )

    const bundles = Array.isArray(fileConfig)
      ? (bundler as Bundler[])
          .map((b) => b.bundles)
          .reduce((b, acc) => {
            return new Set([...acc, ...b])
          }, new Set<any>())
      : (bundler as Bundler).bundles

    for (const bundle of bundles) {
      for (const relative of bundle.keys()) {
        const asset = bundle.get(relative)
        expect(asset && asset.source).toMatchSnapshot(`${title} ${relative}`)
      }
    }
  })
}

snapshot({
  title: 'defaults',
  input: 'index.js',
  cwd: fixture('defaults'),
})

snapshot(
  {
    title: 'banner:true default',
    input: 'index.js',
    cwd: fixture('banner/default'),
  },
  {
    banner: true,
  }
)

snapshot(
  {
    title: 'banner:object',
    input: 'default.js',
    cwd: fixture(),
  },
  {
    banner: {
      author: 'author',
      license: 'GPL',
      name: 'name',
      version: '1.2.3',
    },
  }
)

snapshot(
  {
    title: 'banner:string',
    input: 'default.js',
    cwd: fixture(),
  },
  {
    banner: 'woot',
  }
)

snapshot(
  {
    title: 'exclude file',
    input: 'index.js',
    cwd: fixture('exclude-file'),
  },
  {
    externals: ['./foo.js'],
  }
)

snapshot(
  {
    title: 'extendOptions',
    input: ['foo.js', 'bar.js'],
    cwd: fixture('extend-options'),
  },
  {
    output: {
      format: ['umd', 'umd-min', 'cjs'],
    },
    extendConfig(config, { format }) {
      if (format === 'umd') {
        config.output.moduleName = 'umd'
      }
      if (format.endsWith('-min')) {
        config.output.moduleName = 'min'
      }
      return config
    },
  }
)

snapshot(
  {
    title: 'bundle-node-modules',
    input: 'index.js',
    cwd: fixture('bundle-node-modules'),
  },
  {
    bundleNodeModules: true,
  }
)

snapshot({
  title: 'async',
  input: 'index.js',
  cwd: fixture('async'),
})

snapshot({
  title: 'babel:with-config',
  input: 'index.js',
  cwd: fixture('babel/with-config'),
})

snapshot(
  {
    title: 'babel:disable-config',
    input: 'index.js',
    cwd: fixture('babel/with-config'),
  },
  {
    babel: {
      babelrc: false,
    },
  }
)

snapshot({
  title: 'babel:object-rest-spread',
  input: 'index.js',
  cwd: fixture('babel/object-rest-spread'),
})

snapshot(
  {
    title: 'uglify',
    input: 'index.js',
    cwd: fixture('uglify'),
  },
  {
    output: {
      format: 'cjs-min',
    },
  }
)

snapshot(
  {
    title: 'inline-certain-modules',
    input: 'index.js',
    cwd: fixture('inline-certain-modules'),
  },
  {
    bundleNodeModules: ['fake-module'],
  }
)

snapshot({
  title: 'vue plugin',
  input: 'component.vue',
  cwd: fixture('vue'),
})

snapshot({
  title: 'Typescript',
  input: 'index.ts',
  cwd: fixture('typescript'),
})

snapshot(
  {
    title: 'custom rollup plugin',
    input: 'index.js',
    cwd: fixture('custom-rollup-plugin'),
  },
  {
    plugins: {
      strip: {
        functions: ['console.log'],
      },
    },
  }
)

snapshot(
  {
    title: 'custom scoped rollup plugin',
    input: 'index.js',
    cwd: fixture('custom-scoped-rollup-plugin'),
  },
  {
    plugins: {
      '@rollup/plugin-strip': {
        functions: ['console.log'],
      },
    },
  }
)

snapshot(
  {
    title: '`@rollup/plugin-replace` can accepts custom options',
    input: 'index.js',
    cwd: fixture('custom-scoped-rollup-plugin/replace'),
  },
  {
    plugins: {
      replace: {
        'process.env.NODE_ENV': JSON.stringify('production'),
      },
    },
  }
)

snapshot(
  {
    title: 'target:browser',
    input: 'index.js',
    cwd: fixture('target/browser'),
  },
  {
    output: {
      target: 'browser',
    },
  }
)

snapshot(
  {
    title: 'umd and iife should drop process.env.NODE_ENV',
    input: 'index.js',
    cwd: fixture('format'),
  },
  {
    output: {
      format: ['umd', 'umd-min', 'iife', 'iife-min'],
      moduleName: 'dropNodeEnv',
    },
  }
)

snapshot(
  {
    title: 'umd and iife should preserve existing env.NODE_ENV',
    input: 'index.js',
    cwd: fixture('format'),
  },
  {
    output: {
      format: ['umd', 'umd-min', 'iife', 'iife-min'],
      moduleName: 'dropNodeEnv',
    },
    env: { NODE_ENV: 'production' },
  }
)

snapshot(
  {
    title: 'multiple bundle instance',
    input: 'index.js',
    cwd: fixture('defaults'),
  },
  {},
  [
    {
      output: {
        format: ['umd'],
        moduleName: 'ins1',
      },
    },
    {
      output: {
        format: ['umd-min'],
        moduleName: 'ins2',
      },
    },
  ]
)
