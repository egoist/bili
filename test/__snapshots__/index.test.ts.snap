// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`@rollup/plugin-replace\` can accepts custom options: \`@rollup/plugin-replace\` can accepts custom options dist/index.js 1`] = `
"'use strict';

{
  console.log('prod');
}

let a = 27;

{
  a--;
}

console.log(a);
"
`;

exports[`Typescript: Typescript dist/index.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function greeter(person) {
  return 'Hello, ' + person.firstName + ' ' + person.lastName;
}

let user = {
  firstName: 'Jane',
  lastName: 'User'
};
let s = greeter(user);

exports.s = s;
"
`;

exports[`async: async dist/index.js 1`] = `
"'use strict';

function _async(f) {
  return function () {
    for (var args = [], i = 0; i < arguments.length; i++) {
      args[i] = arguments[i];
    }

    try {
      return Promise.resolve(f.apply(this, args));
    } catch (e) {
      return Promise.reject(e);
    }
  };
}

const a = {
  a: 1
};
var index = _async(function () {
  return Object.assign({}, a, {
    b
  });
});

module.exports = index;
"
`;

exports[`babel:disable-config: babel:disable-config dist/index.js 1`] = `
"'use strict';

class index {}

module.exports = index;
"
`;

exports[`babel:object-rest-spread: babel:object-rest-spread dist/index.js 1`] = `
"'use strict';

var index = Object.assign({}, a);

module.exports = index;
"
`;

exports[`babel:with-config: babel:with-config dist/index.js 1`] = `
"'use strict';

class index {}

module.exports = index;
"
`;

exports[`banner:object: banner:object dist/default.js 1`] = `
"/*!
 * name v1.2.3
 * (c) author
 * Released under the GPL License.
 */
'use strict';

var _default = {};

module.exports = _default;
"
`;

exports[`banner:string: banner:string dist/default.js 1`] = `
"woot
'use strict';

var _default = {};

module.exports = _default;
"
`;

exports[`banner:true default: banner:true default dist/index.js 1`] = `
"/*!
 * foo v0.0.0
 * (c) name <email@email.com>
 * Released under the MIT License.
 */
'use strict';

var index = 42;

module.exports = index;
"
`;

exports[`bundle-node-modules: bundle-node-modules dist/index.js 1`] = `
"'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var all = _interopDefault(require('some-package-that-does-not-exists'));

var pokemon = ['pikachu'];

var index = {
  all,
  pokemon
};

module.exports = index;
"
`;

exports[`custom rollup plugin: custom rollup plugin dist/index.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const tt = 'is there anything?';

function afun() {
  return tt;
}

function another() {
  return afun();
}

exports.another = another;
"
`;

exports[`custom scoped rollup plugin (shorthand): custom scoped rollup plugin (shorthand) dist/index.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const tt = 'is there anything?';

function afun() {
  return tt;
}

function another() {
  return afun();
}

exports.another = another;
"
`;

exports[`custom scoped rollup plugin: custom scoped rollup plugin dist/index.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const tt = 'is there anything?';

function afun() {
  return tt;
}

function another() {
  return afun();
}

exports.another = another;
"
`;

exports[`defaults: defaults dist/index.js 1`] = `
"'use strict';

var index = 42;

module.exports = index;
"
`;

exports[`exclude file: exclude file dist/index.js 1`] = `
"'use strict';

var foo = 'foo';

module.exports = foo;
"
`;

exports[`extendOptions: extendOptions dist/bar.js 1`] = `
"'use strict';

var bar = 'bar';

module.exports = bar;
"
`;

exports[`extendOptions: extendOptions dist/bar.umd.js 1`] = `
"(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
	typeof define === 'function' && define.amd ? define(factory) :
	(global = global || self, global.umd = factory());
}(this, (function () { 'use strict';

	var bar = 'bar';

	return bar;

})));
"
`;

exports[`extendOptions: extendOptions dist/bar.umd.min.js 1`] = `
"!function(e,n){\\"object\\"==typeof exports&&\\"undefined\\"!=typeof module?module.exports=n():\\"function\\"==typeof define&&define.amd?define(n):(e=e||self).min=n()}(this,(function(){\\"use strict\\";return\\"bar\\"}));
"
`;

exports[`extendOptions: extendOptions dist/foo.js 1`] = `
"'use strict';

var foo = 'foo';

module.exports = foo;
"
`;

exports[`extendOptions: extendOptions dist/foo.umd.js 1`] = `
"(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
	typeof define === 'function' && define.amd ? define(factory) :
	(global = global || self, global.umd = factory());
}(this, (function () { 'use strict';

	var foo = 'foo';

	return foo;

})));
"
`;

exports[`extendOptions: extendOptions dist/foo.umd.min.js 1`] = `
"!function(e,n){\\"object\\"==typeof exports&&\\"undefined\\"!=typeof module?module.exports=n():\\"function\\"==typeof define&&define.amd?define(n):(e=e||self).min=n()}(this,(function(){\\"use strict\\";return\\"foo\\"}));
"
`;

exports[`inline-certain-modules: inline-certain-modules dist/index.js 1`] = `
"'use strict';

require('another-module');

var fakeModule = a => a;

var index = fakeModule(1);

module.exports = index;
"
`;

exports[`target:browser: target:browser dist/index.js 1`] = `
"'use strict';

var foo = 'browser';

module.exports = foo;
"
`;

exports[`uglify: uglify dist/index.min.js 1`] = `
"\\"use strict\\";Object.defineProperty(exports,\\"__esModule\\",{value:!0});exports.a=1;
"
`;

exports[`umd and iife should drop process.env.NODE_ENV: umd and iife should drop process.env.NODE_ENV dist/index.iife.js 1`] = `
"var dropNodeEnv = (function (exports) {
  'use strict';

  exports.a = '';
  exports.b = '';

  {
    exports.a = 'live in dev';
  }

  {
    exports.b = 'live in dev';
  }

  return exports;

}({}));
"
`;

exports[`umd and iife should drop process.env.NODE_ENV: umd and iife should drop process.env.NODE_ENV dist/index.iife.min.js 1`] = `
"var dropNodeEnv=function(r){\\"use strict\\";return r.a=\\"\\",r.b=\\"\\",r.a=\\"live in prod\\",r.b=\\"live in prod\\",r}({});
"
`;

exports[`umd and iife should drop process.env.NODE_ENV: umd and iife should drop process.env.NODE_ENV dist/index.umd.js 1`] = `
"(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
  typeof define === 'function' && define.amd ? define(['exports'], factory) :
  (global = global || self, factory(global.dropNodeEnv = {}));
}(this, (function (exports) { 'use strict';

  exports.a = '';
  exports.b = '';

  {
    exports.a = 'live in dev';
  }

  {
    exports.b = 'live in dev';
  }

  Object.defineProperty(exports, '__esModule', { value: true });

})));
"
`;

exports[`umd and iife should drop process.env.NODE_ENV: umd and iife should drop process.env.NODE_ENV dist/index.umd.min.js 1`] = `
"!function(e,o){\\"object\\"==typeof exports&&\\"undefined\\"!=typeof module?o(exports):\\"function\\"==typeof define&&define.amd?define([\\"exports\\"],o):o((e=e||self).dropNodeEnv={})}(this,(function(e){\\"use strict\\";e.a=\\"\\",e.b=\\"\\",e.a=\\"live in prod\\",e.b=\\"live in prod\\",Object.defineProperty(e,\\"__esModule\\",{value:!0})}));
"
`;

exports[`umd and iife should preserve existing env.NODE_ENV: umd and iife should preserve existing env.NODE_ENV dist/index.iife.js 1`] = `
"var dropNodeEnv = (function (exports) {
  'use strict';

  exports.a = '';
  exports.b = '';

  {
    exports.a = 'live in prod';
  }

  {
    exports.b = 'live in prod';
  }

  return exports;

}({}));
"
`;

exports[`umd and iife should preserve existing env.NODE_ENV: umd and iife should preserve existing env.NODE_ENV dist/index.iife.min.js 1`] = `
"var dropNodeEnv=function(r){\\"use strict\\";return r.a=\\"\\",r.b=\\"\\",r.a=\\"live in prod\\",r.b=\\"live in prod\\",r}({});
"
`;

exports[`umd and iife should preserve existing env.NODE_ENV: umd and iife should preserve existing env.NODE_ENV dist/index.umd.js 1`] = `
"(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
  typeof define === 'function' && define.amd ? define(['exports'], factory) :
  (global = global || self, factory(global.dropNodeEnv = {}));
}(this, (function (exports) { 'use strict';

  exports.a = '';
  exports.b = '';

  {
    exports.a = 'live in prod';
  }

  {
    exports.b = 'live in prod';
  }

  Object.defineProperty(exports, '__esModule', { value: true });

})));
"
`;

exports[`umd and iife should preserve existing env.NODE_ENV: umd and iife should preserve existing env.NODE_ENV dist/index.umd.min.js 1`] = `
"!function(e,o){\\"object\\"==typeof exports&&\\"undefined\\"!=typeof module?o(exports):\\"function\\"==typeof define&&define.amd?define([\\"exports\\"],o):o((e=e||self).dropNodeEnv={})}(this,(function(e){\\"use strict\\";e.a=\\"\\",e.b=\\"\\",e.a=\\"live in prod\\",e.b=\\"live in prod\\",Object.defineProperty(e,\\"__esModule\\",{value:!0})}));
"
`;

exports[`vue plugin: vue plugin dist/component.css 1`] = `
"
.main-content[data-v-5fccef39]{margin:0 auto;font-family:sans-serif;background-color:#ccf7e2;padding:10px;border-radius:5px;max-width:500px
}"
`;

exports[`vue plugin: vue plugin dist/component.js 1`] = `
"'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var __vue_normalize__ = _interopDefault(require('vue-runtime-helpers/dist/normalize-component.js'));

//
//
//
//
//
//
//
var script = {
  name: 'main-content',
  data: () => ({
    name: 'Bili !'
  })
};

/* script */
const __vue_script__ = script; // For security concerns, we use only base name in production mode. See https://github.com/vuejs/rollup-plugin-vue/issues/258

script.__file = \\"component.vue\\";
/* template */

var __vue_render__ = function () {
  var _vm = this;

  var _h = _vm.$createElement;

  var _c = _vm._self._c || _h;

  return _c('div', {
    staticClass: \\"main-content\\"
  }, [_c('h1', [_vm._v(\\"This is your first component in Vue\\")]), _vm._v(\\" \\"), _c('h3', [_vm._v(_vm._s(_vm.name))])]);
};

var __vue_staticRenderFns__ = [];
/* style */

const __vue_inject_styles__ = undefined;
/* scoped */

const __vue_scope_id__ = \\"data-v-5fccef39\\";
/* module identifier */

const __vue_module_identifier__ = undefined;
/* functional template */

const __vue_is_functional_template__ = false;
/* style inject */

/* style inject SSR */

var component = __vue_normalize__({
  render: __vue_render__,
  staticRenderFns: __vue_staticRenderFns__
}, __vue_inject_styles__, __vue_script__, __vue_scope_id__, __vue_is_functional_template__, __vue_module_identifier__, undefined, undefined);

module.exports = component;
"
`;
